src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs

```diff
@@ -832,22 +833,60 @@ public static int Sum<T>(this ImmutableArray<T> items, Func<T, int> selector)
             return sum;
         }
 
-        internal static Dictionary<K, ImmutableArray<T>> ToDictionary<K, T>(this ImmutableArray<T> items, Func<T, K> keySelector, IEqualityComparer<K>? comparer = null)
-            where K : notnull
-            where T : notnull
+        public static Dictionary<TKey, ImmutableArray<TSource>> ToDictionary<TSource, TKey>(this ImmutableArray<TSource> items, Func<TSource, TKey> keySelector)
+            where TKey : notnull
+            where TSource : class
+        {
+            return ToDictionary<TSource, TSource, TKey, VoidResult, TSource>(items, static (x, _) => x, keySelector, data: default, downcastUnused: null);
+        }
+
+        /// <summary>
+        /// Produces a dictionary from a list of <paramref name="items"/> mapping from <typeparamref name="TKey"/>s to
+        /// an array of <typeparamref name="TResult"/>s.  The values of the dictionary are produced using <paramref
+        /// name="selector"/> on each item in <paramref name="items"/>.  They keys of the dictionary are producing using
+        /// <paramref name="keySelector"/> on those values.  Values with the same key are grouped into an <see
+        /// cref="ImmutableArray{TResult}"/>.  All arrays returned will always have at least one element in them.
+        /// <para/>
+        /// Clients of this API can ask for special behavior if all values are also an instance of <typeparamref
+        /// name="TDowncast"/>.  In that event an <see cref="ImmutableArray{TDowncast}"/> will be created instead of an
+        /// <see cref="ImmutableArray{TResult}"/>.  This allows clients to directly cast to that type avoiding an
+        /// unnecessary copy when passing around to strongly typed helpers that want to use the derived type.
+        /// </summary>
+        public static Dictionary<TKey, ImmutableArray<TResult>> ToDictionary<TSource, TResult, TKey, TData, TDowncast>(
```

Comment:
>public static Dictionary<TKey, ImmutableArray<TResult>> ToDictionary<TSource, TResult, TKey, TData, TDowncast>( [](http://example.com/codeflow?start=8&length=111)

This doesn't strike me as commonly used helper and I would prefer leaving the code where it was without moving it around to much. We can discuss offline.